spec/models/ability_spec.rb

```diff
@@ -89,11 +89,102 @@
       let(:other_self_organized_track) { create(:track, :self_organized) }
 
       it{ should be_able_to(:manage, user) }
-
       it{ should be_able_to(:manage, registration_public) }
       it{ should be_able_to(:manage, registration_not_public) }
-      it{ should_not be_able_to(:new, Registration.new(conference_id: conference_with_closed_registration.id))}
-      it{ should_not be_able_to(:create, Registration.new(conference_id: conference_with_closed_registration.id))}
+
+      # Test for user can register or not
+      context 'when user is not a speaker with event confirmed' do
+        let(:conference) { create(:conference) }
+
+        context 'when the registration is closed' do
+          let(:conference_with_closed_registration) { create(:conference) }
+
+          it{ should_not be_able_to(:new, Registration.new(conference: conference_with_closed_registration)) }
+          it{ should_not be_able_to(:create, Registration.new(conference: conference_with_closed_registration)) }
+        end
+
+        context 'when the registration period is not set' do
+          it{ should_not be_able_to(:new, Registration.new(conference: conference)) }
+          it{ should_not be_able_to(:create, Registration.new(conference: conference)) }
+        end
+
+        context 'when user have already registered' do
+          let(:user_registration) { create(:registration, conference: conference, user: user) }
+
+          it{ should_not be_able_to(:new, Registration.new(conference_id: user_registration.conference_id)) }
+          it{ should_not be_able_to(:create, Registration.new(conference_id: user_registration.conference_id)) }
+        end
+
+        context 'when user have not registered and registrations are open with conference_limit_not_exceeded' do
```

Comment:
The hidden not in conference_limit_not_exceeded is not very obvious. 
`registration_limit_not_exceeded` derives from the model function that checks this https://github.com/openSUSE/osem/blob/master/app/models/conference.rb#L712, changing this too much, and adding text makes it harder to read in my opinion. 
This can benefit from a small re-write.